lib/iris/fileformats/netcdf/_thread_safe_nc.py

```diff
@@ -35,22 +35,42 @@ class _ThreadSafeWrapper(ABC):
     the C-layer.
     """
 
+    # Note: this is only used to create a "contained" from passed args.
     CONTAINED_CLASS = NotImplemented
+    # Note: this defines how we identify/check that a contained is of the expected type
+    # (in a duck-type way).
+    _DUCKTYPE_CHECK_PROPERTIES: typing.List[str] = [NotImplemented]
 
     # Allows easy type checking, avoiding difficulties with isinstance and mocking.
     THREAD_SAFE_FLAG = True
 
+    @classmethod
+    def is_contained_type(cls, instance):
+        return all(
+            hasattr(instance, attr) for attr in cls._DUCKTYPE_CHECK_PROPERTIES
+        )
+
     @classmethod
     def _from_existing(cls, instance):
         """Pass an existing instance to __init__, where it is contained."""
-        assert isinstance(instance, cls.CONTAINED_CLASS)
+        assert cls.is_contained_type(instance)
         return cls(instance)
 
     def __init__(self, *args, **kwargs):
         """Contain an existing instance, or generate a new one from arguments."""
-        if isinstance(args[0], self.CONTAINED_CLASS):
+        if len(args) > 0 and self.is_contained_type(args[0]):
+            # Passed a contained-type object : Wrap ourself around that.
             instance = args[0]
+            # Explicitly ban "wrapping a wrapper".
+            if hasattr(instance, "THREAD_SAFE_FLAG"):
+                msg = (
+                    "Cannot create {cls} containing an existing {cls}.".format(
+                        cls=self.__class__.__name__
+                    )
+                )
+                raise ValueError(msg)
```

Comment:
IMO an `assert` probably would have been enough for this sort of developer-level stuff, and be a bit more readable. I could be persuaded otherwise.

Reply:
This is good.  I was going to add a test -- but that's definitely overkill !!